src/jit/assertionprop.cpp

```diff
@@ -2212,12 +2212,19 @@ GenTreePtr Compiler::optVNConstantPropOnTree(BasicBlock* block, GenTreePtr stmt,
     case TYP_LONG:
         {
             INT64 value = vnStore->ConstantValue<INT64>(vnCns);
-#ifdef _TARGET_64BIT_           
+#ifdef _TARGET_64BIT_
             if (vnStore->IsVNHandle(vnCns))
             {
-                newTree = gtNewIconHandleNode(value, vnStore->GetHandleFlags(vnCns));
-                newTree->gtVNPair = ValueNumPair(vnLib, vnCns);
-                newTree = optPrepareTreeForReplacement(tree, newTree);
+                // Don't perform constant folding that involves a handle that needs
+                // to be recorded as a relocation with the VM.
+#ifdef RELOC_SUPPORT
+                if (!opts.compReloc)
+#endif
+                {
+                    newTree = gtNewIconHandleNode(value, vnStore->GetHandleFlags(vnCns));
+                    newTree->gtVNPair = ValueNumPair(vnLib, vnCns);
+                    newTree = optPrepareTreeForReplacement(tree, newTree);
+                }
             }
```

Comment:
IMO, we should definitely avoid this #ifdef clutter that appears not once but twice.  What I don't understand is why we need to make IsVNHandle() check twice due to refactoring that i suggested earlier?

```
bool ValueNumStore::IsRelocatableVNHandle(Compiler* comp, valueNum vnCns) const
{
#ifdef RELOC_SUPPORT
    if (comp->opts.compReloc &&  IsVNHandle(vnCns))
    {
         var_types type = TypeOfVN(vnCNS);
  #ifdef _TARGET_64BIT_
         return  (type == TYP_LONG || type == TYP_I_IMPL) 
  #else
         return  (type == TYP_INT || type == TYP_I_IMPL) 
  #endif
    }
#endif

  return false;
}

lines 2215 to 2227 would reduce to

if (!vnStore->IsRelocatableVNHandle(this, vnCns))
{
    newTree = gtNewIconHandleNode(value, vnStore->GetHandleFlags(vnCns));
    newTree->gtVNPair = ValueNumPair(vnLib, vnCns);
    newTree = optPrepareTreeForReplacement(tree, newTree);
}
```